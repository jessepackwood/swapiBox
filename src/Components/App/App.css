.App {
  text-align: center;
  display: flex;
  background-image: url(https://i.ytimg.com/vi/voWpbz1De_M/maxresdefault.jpg);
  margin: 0;
  padding: 0;
  width: 100%;
  min-height: 100vh; }
