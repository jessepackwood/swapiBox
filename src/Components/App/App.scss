.App {
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  background-image: url(https://i.ytimg.com/vi/voWpbz1De_M/maxresdefault.jpg);
  margin: 0;
  padding: 0;
  width: 100%;
  min-height: 100vh;
  font-family: 'Helvetica';
}

.App-title {
  font-size: 57pt;
  letter-spacing: 0.5em;
  color: #feda4a;
  font-family: 'starjedi';
}

.App-container {
  width: 100%
}




