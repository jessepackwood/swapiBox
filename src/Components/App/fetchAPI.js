export const fetchFilm = async () => {
  try {
    const randomFilm = Math.floor(Math.random() * (7) + 1);
    const filmCrawl = await fetch(`https://swapi.co/api/films/${randomFilm}/`);
    const filmData = await filmCrawl.json();
    return Object.assign(
      {},
      {title: filmData.title},
      {episode: filmData.episode_id},
      {text: filmData.opening_crawl}
    );
  } catch (type) {
    return Error('Fetch failed');
  }
};

export const fetchPeople = async () => {
  try {
    const fetchPeople = await fetch('https://swapi.co/api/people/');
    const peopleData = await fetchPeople.json();
    const people = await fetchHomeworldSpecies(peopleData.results);
    return people;
  } catch (type) {
    return Error('Fetch failed');
  }
};

export const fetchHomeworldSpecies = (peopleData) => {
  try {
    const unresolvedPromises = peopleData.map( async (person) => {
      let homeworldFetch = await fetch(person.homeworld);
      let homeworldData = await homeworldFetch.json();
      let speciesFetch = await fetch(person.species);
      let speciesData = await speciesFetch.json();
      return {
        name: person.name,
        info: {
          homeworld: homeworldData.name,
          species: speciesData.classification,
          language: speciesData.language,
          population: homeworldData.population
        }
      };
    });
    return Promise.all(unresolvedPromises);
  } catch (type) {
    return Error('Fetch failed');
  }
};

export const fetchPlanets = async () => {
  try {
    const fetchPlanets = await fetch('https://swapi.co/api/planets/');
    const planetResponse = await fetchPlanets.json();
    const planets = await fetchPlanetsData(planetResponse.results);
    return planets;
  } catch (type) {
    return Error('Fetch failed');
  }
};

export const fetchPlanetsData = (planets) => {
  try {
    const planetsPromises = planets.map( async (planet) => {
      const residentsPromises = planet.residents.map( async (resident) => { 
        const residentData = await fetch(resident);
        const residentObject = await residentData.json();
        return residentObject.name;
      });
      let residentNames = await Promise.all(residentsPromises);
      residentNames = residentNames.length ? residentNames : 'None';

      return {
        name: planet.name,
        info: {
          terrain: planet.terrain,
          population: planet.population,
          climate: planet.climate,
          residents: residentNames
        }
      };
    });

    return Promise.all(planetsPromises);
  } catch (type) {
    return Error('Fetch failed');
  }
};

export const fetchVehicles = async () => {
  try {
    const fetchvehicles = await fetch('https://swapi.co/api/vehicles/');
    const vehicleObj = await fetchvehicles.json();
    const vehicles = vehicleObj.results.map( vehicle => {
      return {
        name: vehicle.name,
        info: {
          model: vehicle.model,
          class: vehicle.vehicle_class,
          passengers: vehicle.passengers
        }
      };
    });
    return vehicles;
  } catch (type) {
    return Error('Fetch failed');
  }
};
