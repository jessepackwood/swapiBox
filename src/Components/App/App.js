import React, { Component } from 'react';
import ScrollText from '../ScrollText/ScrollText';
import CardContainer from '../CardContainer/CardContainer';
import {fetchFilm} from './fetchAPI';
import Header from '../Header/Header';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      display: '',
      favorites: [],
      film: {},
      page: 1,
      people: [],
      planets: [],
      vehicles: []
    };
  }

  async componentDidMount() {
    const film = await fetchFilm();
    this.setState({film});
  }

  setDisplay = async (type, fetchFunction) => {
    let typeState = this.state[type];
    let response;
    if (typeState.length) {
      this.setState({display: type});
    } else {
      response = await fetchFunction();
      this.setState({[type]: response, display: type});
    }
  }
    
  toggleFavorite = (cardObject) => {
    let favorites = this.state.favorites;

    const currentFavorites = 
      favorites.find(favorite => favorite.title === cardObject.title);
    if (!currentFavorites) {
      favorites.push(cardObject);
    } else {
      favorites = favorites.filter(favorite =>
        favorite.title !== cardObject.title);
    }
    this.setState({favorites});
  };

  showFavorites = () => {
    this.setState({display: 'favorites'});
  };

  onSkip = () => {
    this.setState({page: 2});
  };

  render() {
    return (
      <div className="App">
        <h1 className="App-title">SwapiBox</h1>
        <div className='App-container'>
          {this.state.page === 1 && 
            <ScrollText 
              className='scroll-text'
              film={this.state.film}
              onSkip={this.onSkip}
            />}
          {this.state.page === 2 && 
            <div>
              <Header
                buttonText=''
                setDisplay={this.setDisplay}
                showFavorites={this.showFavorites}
                favoriteCount={this.state.favorites.length}
              />
              <CardContainer 
                people={this.state.people}
                planets={this.state.planets}
                vehicles={this.state.vehicles} 
                display={this.state.display}
                toggleFavorite={this.toggleFavorite}
                favorites={this.state.favorites}
              />
            </div>
          }
        </div>
      </div>
    );
  }
}

export default App;
