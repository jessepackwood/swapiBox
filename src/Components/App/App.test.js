import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme'
import './fetchAPI'
import App from './App';

jest.mock('./fetchAPI.js');

describe('App test', () => {
  beforeEach(() => {
    window. fetch = 
    jest.fn().mockImplementation(() => Promise.resolve({
    status: 200,
    json: () => Promise.resolve(
      {opening_crawl: 'Star Warssss',

       title: 'Return of the Jedi',
     }
    )
    }))
  });
it('should load scrollText on mount', async () => {
  // const mockFetch = jest.fn().mockImplementation(() => Promise.resolve({
  //   status: 200,
  //   json: () => Promise.resolve(
  //     {opening_crawl: 'Star Warssss',

  //      title: 'Return of the Jedi',
  //    }
  //   )
  // }))
  // window.fetch = mockFetch
  const app = await shallow(<App />, {disableLifeCycleMethods: true});
  await app.update()
  expect(app.state('film').opening_crawl).toEqual('Star Warssss')
  })

  it('should load people when the people button is clicked', () => {
    
  })
});