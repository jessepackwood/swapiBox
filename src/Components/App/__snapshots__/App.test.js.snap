// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test Should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="App-title"
>
          SwapiBox
</h1>,
        <div
          className="App-container"
>
          <ScrollText
                    className="scroll-text"
                    film={Object {}}
                    onSkip={[Function]}
          />
</div>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "SwapiBox",
          "className": "App-title",
        },
        "ref": null,
        "rendered": "SwapiBox",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ScrollText
              className="scroll-text"
              film={Object {}}
              onSkip={[Function]}
/>,
            false,
          ],
          "className": "App-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "scroll-text",
              "film": Object {},
              "onSkip": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="App-title"
>
            SwapiBox
</h1>,
          <div
            className="App-container"
>
            <ScrollText
                        className="scroll-text"
                        film={Object {}}
                        onSkip={[Function]}
            />
</div>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SwapiBox",
            "className": "App-title",
          },
          "ref": null,
          "rendered": "SwapiBox",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ScrollText
                className="scroll-text"
                film={Object {}}
                onSkip={[Function]}
/>,
              false,
            ],
            "className": "App-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "scroll-text",
                "film": Object {},
                "onSkip": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
